sys/net/gnrc/link_layer/lwmac/lwmac_internal.c

```diff
@@ -0,0 +1,538 @@
+/*
+ * Copyright (C) 2015 Daniel Krebs
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @{
+ * @ingroup     net_lwmac
+ * @file
+ * @brief       Internal functions of LWMAC
+ *
+ * @author      Daniel Krebs <github@daniel-krebs.net>
+ * @}
+ */
+
+#include <stdbool.h>
+#include <periph/rtt.h>
+#include <net/gnrc.h>
+#include <net/gnrc/lwmac/lwmac.h>
+#include <net/gnrc/lwmac/packet_queue.h>
+
+#include "include/lwmac_internal.h"
+#include "include/lwmac_types.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+/******************************************************************************/
+
+int _get_dest_address(gnrc_pktsnip_t* pkt, uint8_t* pointer_to_addr[])
+{
+    int res;
+    gnrc_netif_hdr_t* netif_hdr;
+
+    if(!pkt)
+        return -ENODEV;
+
+    netif_hdr = (gnrc_netif_hdr_t*) pkt->data;
+    if( (res = netif_hdr->dst_l2addr_len) <= 0)
+        return -ENOENT;
+
+    *pointer_to_addr = gnrc_netif_hdr_get_dst_addr(netif_hdr);
+    return res;
+}
+
+/******************************************************************************/
+
+/* Find a payload based on it's protocol type */
+void* _gnrc_pktbuf_find(gnrc_pktsnip_t* pkt, gnrc_nettype_t type)
+{
+    while(pkt != NULL)
+    {
+        if(pkt->type == type) {
+            return pkt->data;
+        }
+        pkt = pkt->next;
+    }
+    return NULL;
+}
+
+/******************************************************************************/
+
+int _find_neighbour(lwmac_t* lwmac, uint8_t* dst_addr, int addr_len)
+{
+    lwmac_tx_neighbour_t* neighbours = lwmac->tx.neighbours;
+
+    for(int i = 0; i < LWMAC_NEIGHBOUR_COUNT; i++) {
+        if(neighbours[i].l2_addr.len == addr_len) {
+            if(memcmp(&(neighbours[i].l2_addr.addr), dst_addr, addr_len) == 0) {
+                return i;
+            }
+        }
+    }
+    return -1;
+}
+
+/******************************************************************************/
+
+/* Free first empty queue that is not active */
+int _free_neighbour(lwmac_t* lwmac)
+{
+    lwmac_tx_neighbour_t* neighbours = lwmac->tx.neighbours;
+
+    for(int i = 0; i < LWMAC_NEIGHBOUR_COUNT; i++) {
```

Comment:
I think, for ```_free_neighbour``` function, the parameter ```i``` in the ```for``` loop here should start with ```1```, but not ```0```. If ```i``` start with 0, there is a possibility that, when ```_free_neighbour``` was called in ```_queue_tx_packet``` to free a neighbor, then the code will first check ```neighbours[0]``` which is only for broadcasting packets.  And, in case ```neighbours[i]``` 's queue length is also ```0```, then ```neighbours[0]``` will be freed, which is unwanted.

I run a test for ensuring this issue, I set the ```LWMAC_NEIGHBOUR_COUNT``` to 2, and try to push two packets for two different receivers, and then, I did found that ```neighbours[0]``` was freed for non-broadcasting purpose.

```
Welcome to pyterm!
Type '/exit' to exit.
txtsnd 4 11:22 4545
2016-06-07 12:01:58,840 - INFO # txtsnd 4 11:22 4545
2016-06-07 12:01:58,845 - INFO # Shuguo: the current find neighbour 1 's queue-length is 1. 
txtsnd 4 33:44 4545
2016-06-07 12:02:05,720 - INFO # > txtsnd 4 33:44 4545
2016-06-07 12:02:05,726 - INFO # Shuguo: the current find neighbour 0 's queue-length is 1. 
```